### PR TITLE
Revise RFC1, 2 & 3 to replace status-code

### LINKED ISSUE #23: ⚗️Review BAM/COMIT protocols post RFC1-3
Now that we have specified two RFCs which use BAM it would be good to investigate how well it solved the problems it was designed to and if are we using it in the right way.

A specific topic that must be addressed the design of and our use of status codes, particularly with regards to the "reason" headers. We had a discussion of this and there were multiple ideas that are simply here as a reminder.

1. We should use different status codes to represent every type of error including those that are related to the swap protocol parameters. e.g. `timeouts-too-tight`. We could partition them according to the "layer" they occurred to create some separation.
2. We should introduce a new type of status code e.g. `NF**` "negotiation failed" to indicate the proposed action was understood but was rejected.
3. The scope of RFC01 might be too big. Maybe some stuff should move into a more domain specific RFC.

Note this ticket isn't meant to imply that there is something wrong with the way things are now.

DoD:
- organise discussions around the topic
- organise decision

Blocked by coblox/COMIT-RFCs#7

### PR DESCRIPTION
**Pull request by thomaseizinger**
The changes proposed in this RFC were mainly driven by the need to harmonize who errors are communicated to the other party.
The basic idea is to split errors up into two categories:

1. errors related to lowest layer of the protocol, e.g. BAM

These are now communicated through an ERROR frame and are usually unrecoverable.

2. errors related to a particular message type (like SWAP)

RFC2 now introduces its own header to encode the status information called `negotiation_result`. This keeps the scope of the feature smaller and is thus less awkward to reason about and implement.

Resolves #23.

### CODE REVIEW

File: RFC-002-SWAP.md [hunk 1]
```diff
@@ -10,267 +10,273 @@
...
-### `status`
+### SWAP RESPONSE frame
 
-See [RFC-001](./RFC-001-BAM.md#status-code-families) for more details, including the definition of statuses `XX00-19`.
+#### Definition
 
-RFC-002 reserves statuses 20 to 39 across all families.
-Protocols may define the meaning of statuses 40 and above.
+A frame of type `REQUEST` implies a `RESPONSE`.
+We will refer to the response of a `SWAP REQUEST` as SWAP RESPONSE.
```
Comment (luckysori) [inline RFC-002-SWAP.md]: :wrench: Only place where swap request or swap response is highlighted this way.
```suggestion
We will refer to the response of a SWAP REQUEST as SWAP RESPONSE.
```

File: RFC-002-SWAP.md [hunk 2]
```diff
@@ -10,267 +10,273 @@
...
 
-* `OK20`: Accepted
-* `RE20`: Declined - the Receiver voluntarily turned down the swap request
-* `RE21`: Rejected - the Receiver is not able to proceed with the swap request
+The SWAP RESPONSE serves two purposes:
 
-### `headers`
-#### `reason` (optional)
-The reason why the Receiver Declined or Rejected the swap request.
+1. It contains the result of the negotiation initiated through the SWAP REQUEST
```
Comment (luckysori) [inline RFC-002-SWAP.md]: :upside_down_face: 
```suggestion
1. It contains the result of the negotiation initiated through the SWAP REQUEST.
```

File: RFC-002-SWAP.md [hunk 3]
```diff
@@ -10,267 +10,273 @@
...
+2. It allows the receiving party to communicate information
```
Comment (luckysori) [inline RFC-002-SWAP.md]: :upside_down_face: 
```suggestion
2. It allows the receiving party to communicate additional information.
```

File: RFC-002-SWAP.md [hunk 4]
```diff
@@ -10,267 +10,273 @@
...
-| `unsupported-ledger`   | The Receiver does not support the requested ledger combination.                   | none | |
-| `unavailable-asset`    | The Receiver does not have the given asset or enough of the given asset quantity. | optional, both or none | `alpha_asset`, `beta_asset` |
+#### Body
 
-### `body`
+Together with the the value of the `protocol` header, the `negotiation_result` header determines the `body` of a SWAP RESPONSE.
 
-Refer to the RFC of the given swap protocol.
+In the case of a `successful` negotiation, the `body` of the response will contain the necessary information for the swap to start.
+This entirely depends on the chosen `protocol` and is thereby subject to the `protocol`s RFC to define the `body`.
```
Comment (luckysori) [inline RFC-002-SWAP.md]: :x:
```suggestion
This entirely depends on the chosen `protocol` and is thereby subject to the `protocol`'s RFC to define the `body`.
```

File: RFC-002-SWAP.md [hunk 5]
```diff
@@ -10,267 +10,273 @@
...
+In the case of a `failed` negotiation, the `body` of the response MAY contain an object of type `NegotiationError`.
+A `NegotiationError` is an object with a two properties:
 
-# Examples
+- `reason`: A string, acting as the identifier for this `NegotiationError`.
```
Comment (luckysori) [inline RFC-002-SWAP.md]: :x:
```suggestion
- `reason`: a string acting as the identifier for the `NegotiationError`.
```

File: RFC-002-SWAP.md [hunk 6]
```diff
@@ -10,267 +10,273 @@
...
+- `details`: An object containing more information about the error, depending on the given `reason`.
```
Comment (luckysori) [inline RFC-002-SWAP.md]: :upside_down_face: 
```suggestion
- `details`: an object containing more information about the error, depending on the given `reason`.
```

File: RFC-001-BAM.md [hunk 7]
```diff
@@ -166,127 +169,92 @@ Ids MAY be skipped but MUST be ascending.
 
 > `NOTIFICATION` is a reserved type and may be specified at a later point.
 
+#### Error
+
+The `ERROR` frame is used to communicate failures between nodes on the lowest level of this protocol.
+It is purely informational and is meant to fascilitate debugging.
```
Comment (D4nte) [inline RFC-001-BAM.md]:
```suggestion
It is purely informational and is meant to facilitate debugging.
```

File: RFC-001-BAM.md [hunk 8]
```diff
@@ -166,127 +169,92 @@ Ids MAY be skipped but MUST be ascending.
...
+An `ERROR` frame can not be sent pro-actively.
+It's `id` MUST refer to a previously received frame, like a `REQUEST` frame.
```
Comment (D4nte) [inline RFC-001-BAM.md]:
```suggestion
Its `id` MUST refer to a previously received frame, like a `REQUEST` frame.
```

File: RFC-002-SWAP.md [hunk 9]
```diff
@@ -10,267 +10,273 @@
...
+The protocol specified in the `protocol` header is not known to the receiving party.
+
+##### `details`
+
+TBD <!-- List known protocols in details -->
+
+#### `unknown-ledger`
+
+A ledger referenced by the sending party is unknown to the receiving party.
+Note that different networks of the same blockchain are different ledger!
```
Comment (luckysori) [inline RFC-002-SWAP.md]: :wrench: 
```suggestion
Note that different networks of the same blockchain are different ledgers!
```

File: RFC-001-BAM.md [hunk 10]
```diff
@@ -166,127 +169,92 @@ Ids MAY be skipped but MUST be ascending.
...
-- OK00 (read as OK-00)
-- SE00 (read as **S**ender**E**rror-00)
-- RE00 (read as **R**eceiver**E**rror-00)
-
-This protocol reserves the status codes 00 to 19 for all of the families.
-Application protocols that build on top of this transport protocol MUST NOT redefine any status codes within this range.
-However, they are highly encouraged to define status codes started from `20` for their own purposes.
+To avoid more complexity through additional messages, the spec doesn't define a concept of acknowledging successful processing of a `RESPONSE` to the sender.
+Thus, there is no way of signaling to the sender of a `RESPONSE` whether or not it was properly understood.
+Therefore, careful thought should be put into the design of and use of the `MUST understand` variant of a header to make this failure case as rare as possible.
```
Comment (D4nte) [inline RFC-001-BAM.md]: I am not sure I am following. if there is a `MUST understand` that is not understand then the other node will send an `ERROR` frame, right?

How is it related to lack of acknowledgement type frames?

File: RFC-002-SWAP.md [hunk 11]
```diff
@@ -10,267 +10,273 @@
...
+| Header name          | Value                      |
+| -------------------- | -------------------------- |
+| `alpha_ledger`       | `Ledger`                   |
+| `betaledger`         | `Ledger`                   |
+| `alpha_asset`        | `Asset`                    |
+| `beta_asset`         | `Asset`                    |
+| `protocol`           | `SwapProtocol`             |
+| `negotiation_result` | `"successful" OR "failed"` |
+
+### The following `NegotiationError`s
```
Comment (luckysori) [inline RFC-002-SWAP.md]: :upside_down_face: 
```suggestion
### Several instances of `NegotiationError`
```

File: RFC-001-BAM.md [hunk 12]
```diff
@@ -166,127 +169,92 @@ Ids MAY be skipped but MUST be ascending.
...
+In particular, nodes should never send a `RESPONSE` that contains a `MUST understand` header without them having confidence that the receiving node will understand it.
```
Comment (D4nte) [inline RFC-001-BAM.md]: should they? if the other node does not understand it they will get an `ERROR` frame.

With the introduction of the `ERROR` frame a node can feel free to send requests with `MUST understand` headers and if they receive an error, they can try again without such headers.

We are telling the implementator not to do that, but why?

File: RFC-001-BAM.md [hunk 13]
```diff
@@ -166,127 +169,92 @@ Ids MAY be skipped but MUST be ascending.
...
-A sending node MAY infer that the request itself was fine but the receiving end failed to process it due to an internal error which the sending end MUST NOT be held responsible for.
-
-1. RE00 - Internal error
-
-    `RE00` indicates that the node encountered an internal, unexpected error during processing of the request.
+The structure of the `body` of a REQUEST is entirely up to the application protocol that is defined through the REQUEST `type`.
+Similar to HTTP, application protocols MAY include some kind of 'Content-Type' in the headers in order to describe the encoding of the body.
+When designing an application protocol, it is common to overcome the question of whether a specific piece of data should be encoded as a header or be represented in the body.
+The rule of thumb here is that implementations should be able to parse all headers orthogonally.
+Hence, the structure/possible values of one header SHOULD NOT depend on those of other headers.
```
Comment (D4nte) [inline RFC-001-BAM.md]: 🤔 we kind of have that with ledgers and assets (kinda)

File: RFC-003-SWAP-Basic.md [hunk 14]
```diff
@@ -214,88 +201,106 @@ A security model of the protocol and its associated parameters will be included
...
+
+A new section for listing hash functions is added.
+The section is bootstrapped with the `SHA-256` hash function.
 
 ## References
 1. https://en.bitcoin.it/wiki/Atomic_swap
 2. https://tools.ietf.org/html/rfc4634#section-4.1
 
-# Examples
+## Examples
```
Comment (luckysori) [inline RFC-003-SWAP-Basic.md]: :wrench: As in RFC002
```suggestion

Elements not relevant for this RFC or which are subject to later definition are filled in with "...".

```

File: registry.md [hunk 15]
```diff
@@ -83,59 +76,72 @@ And the possible parameters they each may have:
...
+The following is a list of errors that receivers of a SWAP REQUEST may choose to send back to the sender. In order to send a negotiation error, the value of the `negotiation_result` header MUST be set to "failed".
 
-The following is a list of protocols defined in COMIT RFCs for use in the `protocol` header of a SWAP message.
+| `reason`               | Reference                          | Description                                                                                                                                                              | `details`                                           |
+| :--------------------- | ---------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | --------------------------------------------------- |
+| `unsatisfactory-rate`  | [RFC-002](./RFC-002-SWAP.md)       | The rate of `alpha_asset` to `beta_asset` is not satisfactory to the receiver.                                                                                           | None                                                |
```
Comment (luckysori) [inline registry.md]: :upside_down_face: 
```suggestion
| `unsatisfactory-rate`  | [RFC-002](./RFC-002-SWAP.md)       | The rate of `alpha_asset` to `beta_asset` is not satisfactory to the receiver                                                                                            | None                                                |
```

File: registry.md [hunk 16]
```diff
@@ -83,59 +76,72 @@ And the possible parameters they each may have:
...
+| `protocol-unsupported` | [RFC-002](./RFC-002-SWAP.md)       | The protocol specified in the `protocol` header is not known to the receiving party.                                                                                     | None                                                |
```
Comment (luckysori) [inline registry.md]: :upside_down_face: 
```suggestion
| `protocol-unsupported` | [RFC-002](./RFC-002-SWAP.md)       | The protocol specified in the `protocol` header is not known to the receiving party                                                                                      | None                                                |
```

File: registry.md [hunk 17]
```diff
@@ -83,59 +76,72 @@ And the possible parameters they each may have:
...
+| `unknown-ledger`       | [RFC-002](./RFC-002-SWAP.md)       | A ledger referenced by the sending party is unknown to the receiving party.                                                                                              | None                                                |
```
Comment (luckysori) [inline registry.md]: :upside_down_face: 
```suggestion
| `unknown-ledger`       | [RFC-002](./RFC-002-SWAP.md)       | A ledger referenced by the sending party is unknown to the receiving party                                                                                               | None                                                |
```

File: registry.md [hunk 18]
```diff
@@ -83,59 +76,72 @@ And the possible parameters they each may have:
...
+| `unknown-asset`        | [RFC-002](./RFC-002-SWAP.md)       | An asset referenced by the sending party is unknown to the receiving party.                                                                                              | None                                                |
```
Comment (luckysori) [inline registry.md]: :upside_down_face: 
```suggestion
| `unknown-asset`        | [RFC-002](./RFC-002-SWAP.md)       | An asset referenced by the sending party is unknown to the receiving party                                                                                               | None                                                |
```

File: registry.md [hunk 19]
```diff
@@ -83,59 +76,72 @@ And the possible parameters they each may have:
...
+| `timeouts-too-tight`   | [RFC-003](./RFC-003-SWAP-Basic.md) | This indicates to the sender that the difference between `alpha_expiry` and `beta_expiry` is too small and the receiver may accept the swap if they are given more time. | [Details](#details-of-the-timeouts-too-tight-error) |
```
Comment (luckysori) [inline registry.md]: :upside_down_face: 
```suggestion
| `timeouts-too-tight`   | [RFC-003](./RFC-003-SWAP-Basic.md) | This indicates to the sender that the difference between `alpha_expiry` and `beta_expiry` is too small and the receiver may accept the swap if they are given more time  | [Details](#details-of-the-timeouts-too-tight-error) |
```

File: RFC-001-BAM.md [hunk 20]
```diff
@@ -427,16 +394,17 @@ This section defines how headers are encoded in the JSON-based text encoding.
...
 ```json
-"_source_ledger" : {
-    "value": "Bitcoin",
-    "parameters": {
-        "network": "mainnet"
-    }
+"_payment_method" : {
+  "value": "credit-card",
+  "parameters": {
+    "provider": "visa",
```
Comment (D4nte) [inline RFC-001-BAM.md]:
```suggestion
    "provider": "tenx",
```
I am not keen to have too much references to the old financial system in our repo ;)

File: RFC-003-SWAP-Basic.md [hunk 21]
```diff
@@ -103,39 +105,24 @@ How to construct HTLCs based on SHA-256 and other hash functions for particular
...
+| `beta_redeem_identity`  | `β::Identity`       | The identity on β that **B** will be transferred to when the β-HTLC is activated with the correct secret |
+| `secret_hash`           | `hex-encoded-bytes` | The output of calling `hash_function` on the secret                                                      |
 
 If `alpha_expiry` or `beta_expiry` are in the past, implementations SHOULD consider the request to be invalid.
 
-### Swap Response (Accept)
+### SWAP Response
 
-If responding with `OK00`, the responder MUST include the following fields in the response body:
+If responding with "successful" for the `negotiation_result` header, the responder MUST include the following fields in the response body:
```
Comment (D4nte) [inline RFC-003-SWAP-Basic.md]:
```suggestion
If responding with `successful` for the `negotiation_result` header, the responder MUST include the following fields in the response body:
```

File: RFC-003-SWAP-Basic.md [hunk 22]
```diff
@@ -214,88 +201,106 @@ A security model of the protocol and its associated parameters will be included
...
-    }
+  "type": "RESPONSE",
+  "id": 0,
+  "payload": {
+    "headers": {
+      "negotiation_result": "failed"
+    },
+    "body": { 
+      "reason": "timeouts-too-tight",
+      "details": {}
```
Comment (D4nte) [inline RFC-003-SWAP-Basic.md]: Adding `min_time` to the example would be great.

File: RFC-003-SWAP-Basic.md [hunk 23]
```diff
@@ -214,88 +201,106 @@ A security model of the protocol and its associated parameters will be included
...
+### Negotiation error for tight timeouts
+
+A `NegotiationError` with the reason `timeouts-too-tight` is added.
+This indicates to the sender that the difference between `alpha_expiry` and `beta_expiry` is too small and the receiver may accept the swap if they are given more time.
+
+#### `details`
+
+| property name | type   | description                                                                         |
+| :------------ | :----- | :---------------------------------------------------------------------------------- |
+| min_time      | number | The minimum time difference between the HLTCs in seconds that the receiver requires |
```
Comment (D4nte) [inline RFC-003-SWAP-Basic.md]: It is unclear whether this property is mandatory or optional.

File: registry.md [hunk 24]
```diff
@@ -1,68 +1,62 @@
...
 | `testnet` | Bitcoin Core testnet                 |
 | `mainnet` | Bitcoin Core mainnet                 |
 
-
 ### `ethereum` Parameters
 
-| Parameter | Value Type                                    | Description                              |
-|:----------|-----------------------------------------------|------------------------------------------|
+| Parameter | Value Type                                  | Description                              |
+| :-------- | ------------------------------------------- | ---------------------------------------- |
```
Comment (D4nte) [inline registry.md]: It would be great if all this useless auto-reformat were done in another PR to not clutter this PR :)
When `cargo fmt` for markdown?

File: RFC-001-BAM.md [hunk 25]
```diff
@@ -166,127 +169,92 @@ Ids MAY be skipped but MUST be ascending.
...
-
-###### Receiver errors (RE00-RE99)
-
-Responses from the `RE` family indicate that the receiving node encountered an error during process of the request.
-A sending node MAY infer that the request itself was fine but the receiving end failed to process it due to an internal error which the sending end MUST NOT be held responsible for.
-
-1. RE00 - Internal error
-
-    `RE00` indicates that the node encountered an internal, unexpected error during processing of the request.
+The structure of the `body` of a REQUEST is entirely up to the application protocol that is defined through the REQUEST `type`.
```
Comment (LLFourn) [inline RFC-001-BAM.md]: The second part of this sentence muddled up the point of the sentence for me.
```suggestion
The structure of the `body` of a REQUEST is entirely up to the application protocol.
``` 
Maybe then a sentence saying it's "usually a function of the REQUEST `type` and headers"

File: RFC-001-BAM.md (hunk 13 above)
Comment (LLFourn) [reply RFC-001-BAM.md]: Yeah I would remove the `/possible values` bit of this sentence.

File: RFC-002-SWAP.md [hunk 26]
```diff
@@ -10,267 +10,273 @@
...
+#### Headers
 
-#### `protocol`
-A protocol that defines the steps, transactions and communications needed to proceed with an asset exchange.
+To express all the information for an exchange, a SWAP REQUEST MUST include the following headers:
 
-The exact value of this header will be defined in subsequent swap protocol RFCs.
+- `alpha_ledger`: describes the ledger the alpha-asset is tracked on
+- `beta_ledger`: describes the ledger the beta-asset is tracked on
+- `alpha_asset`: describes the asset who's ownership is transferred on the alpha-ledger
```
[the remark this answers is not in the document]
Comment (LLFourn) [reply RFC-002-SWAP.md]: errr it should be `alpha_ledger` though.

File: RFC-001-BAM.md (hunk 12 above)
Comment (thomaseizinger) [reply RFC-001-BAM.md]: The `ERROR` frame can never be sent as a response to a RESPONSE. It can only be sent instead of a RESPONSE.

Comment (D4nte) [reply RFC-001-BAM.md]: ah, gotcha

File: RFC-001-BAM.md [hunk 27]
```diff
@@ -166,127 +169,92 @@ Ids MAY be skipped but MUST be ascending.
 
 > `NOTIFICATION` is a reserved type and may be specified at a later point.
 
+#### Error
+
+The `ERROR` frame is used to communicate failures between nodes on the lowest level of this protocol.
```
[the remark this answers is not in the document]
Comment (LLFourn) [reply RFC-001-BAM.md]: @thomaseizinger I made a branch giving an example of how I'd introduce this concept.  While doing it I noticed it didn't really fit with the way the rest of the document is structured, so I just put it at the end of the document. After writing it, I'm growing in favour of putting things in this form:

https://github.com/comit-network/RFCs/blob/23-LL-bam-channel/RFC-001-BAM.md#channels

Up to you, whether you want to include it in the scope of this PR.

[the remark this answers is not in the document]
Comment (thomaseizinger) [reply RFC-001-BAM.md]: I think, this should be another revision on its own, since it seems like a fairly isolated change.

File: RFC-001-BAM.md (hunk 10 above)
Comment (thomaseizinger) [reply RFC-001-BAM.md]: `ERROR` frames are an alternative to `RESPONSE` frames. There are no "responses" to invalid `RESPONSE` frames.

File: RFC-001-BAM.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (thomaseizinger) [reply RFC-001-BAM.md]: Good catch, thanks! :)

File: RFC-001-BAM.md (hunk 13 above)
Comment (thomaseizinger) [inline RFC-001-BAM.md]:
```suggestion
Hence, the structure of one header SHOULD NOT depend on those of other headers.
```

File: RFC-001-BAM.md [hunk 28]
```diff
@@ -477,17 +445,17 @@ The following is therefore invalid:
 
 ```json
 "invalid_header": {
-    "some_key": "foobar"
+  "some_key": "foobar"
 }
 ```
 
 If a header needs an `object` to express its value, you should resort to the default representation to make it unambigous:
```
Comment (thomaseizinger) [inline RFC-001-BAM.md]:
```suggestion
If a header needs an `object` to express its value, you should resort to the default representation to make it unambiguous:
```

File: RFC-002-SWAP.md [hunk 29]
```diff
@@ -10,267 +10,273 @@
...
-- [`alpha_ledger`](#alpha_ledger)
-- [`beta_asset`](#beta_asset)
-- [`beta_ledger`](#beta_ledger)
-- [`protocol`](#protocol)
+1. A terminology for describing the exchange of assets
+2. A REQUEST/RESPONSE pair of messages for `BAM!` to negotiate such an exchange between two parties
 
 ## Terminology
 
+To describe the elements of an exchange, we are using the terms `Ledger`, `Asset` and `Alpha/Beta`.
```
Comment (thomaseizinger) [inline RFC-002-SWAP.md]:
```suggestion
To describe the elements of an exchange, we use the terms `Ledger`, `Asset`, `Alpha` and `Beta`.
```

File: RFC-002-SWAP.md (hunk 26 above)
Comment (thomaseizinger) [inline RFC-002-SWAP.md]:
```suggestion
- `alpha_asset`: describes the asset who's ownership is transferred on the `alpha_ledger`
```

File: RFC-002-SWAP.md [hunk 30]
```diff
@@ -10,267 +10,273 @@
...
 
-#### `protocol`
-A protocol that defines the steps, transactions and communications needed to proceed with an asset exchange.
+To express all the information for an exchange, a SWAP REQUEST MUST include the following headers:
 
-The exact value of this header will be defined in subsequent swap protocol RFCs.
+- `alpha_ledger`: describes the ledger the alpha-asset is tracked on
+- `beta_ledger`: describes the ledger the beta-asset is tracked on
+- `alpha_asset`: describes the asset who's ownership is transferred on the alpha-ledger
+- `beta_asset`: describes the asset who's ownership is transferred on the beta-ledger
```
Comment (thomaseizinger) [inline RFC-002-SWAP.md]:
```suggestion
- `beta_asset`: describes the asset whose ownership will be transferred on the `beta_ledger`
```

File: RFC-002-SWAP.md [hunk 31]
```diff
@@ -10,267 +10,273 @@
...
-| `protocol-unsupported` | The Receiver does not support the requested protocol.                             | optional | `protocol` |
-| `unsupported-ledger`   | The Receiver does not support the requested ledger combination.                   | none | |
-| `unavailable-asset`    | The Receiver does not have the given asset or enough of the given asset quantity. | optional, both or none | `alpha_asset`, `beta_asset` |
+#### Body
 
-### `body`
+Together with the the value of the `protocol` header, the `negotiation_result` header determines the `body` of a SWAP RESPONSE.
 
-Refer to the RFC of the given swap protocol.
+In the case of a `successful` negotiation, the `body` of the response will contain the necessary information for the swap to start.
```
[the remark this answers is not in the document]
Comment (thomaseizinger) [reply RFC-002-SWAP.md]: What should the header name be then?

- `swap_result`?
- `result`?
- `swap_status`?

File: registry.md (hunk 24 above)
Comment (thomaseizinger) [reply registry.md]: Too late at this stage :see_no_evil:

File: assets/RFC002-parse-response-body.puml [hunk 32]
```diff
@@ -0,0 +1,18 @@
+@startuml
```
[the remark this answers is not in the document]
Comment (thomaseizinger) [reply assets/RFC002-parse-response-body.puml]: I am using the plantuml proxy service for this. It is documented here: https://blog.anoff.io/2018-07-31-diagrams-with-plantuml/#github-integration

The rendered version is embedded in the RFC once it is merged and available on master.

I purposely did not include a rendered version so that it cannot get outdated.

File: RFC-001-BAM.md [hunk 33]
```diff
@@ -166,127 +173,112 @@ Ids MAY be skipped but MUST be ascending.
...
 
-    A response with the status code `SE02` indicates that the sender sent a request with a `type` that was not understood.
+If a node receives a header in the `MUST understand` variant in a `REQUEST` and it does not understand it, it MUST reject the request with an `ERROR` frame of type `unknown-mandatory-header`.
+Headers encoded as `MAY ignore` are ok to be not understood.
+Nodes may simply ignore them as if they were not there.
 
-###### Receiver errors (RE00-RE99)
+To avoid more complexity through additional messages, the spec doesn't define a concept for acknowledging successful processing of a `RESPONSE` to the sender.
+Thus, there is no way of signaling to the sender of a `RESPONSE` whether or not it was properly understood.
+Therefore, careful thought should be put into the design and use of the `MUST understand` variant of a header to make this failure case as rare as possible.
```
[the remark this answers is not in the document]
Comment (thomaseizinger) [reply RFC-001-BAM.md]: Task: Remove the `successful` part in line 266.

File: RFC-001-BAM.md [hunk 34]
```diff
@@ -507,11 +499,29 @@ The following section describes the behaviour in certain failure scenarios.
...
+
+This RFC adds a section "FRAME types" to track the list of available `type`s to be used for `FRAME`s.
+The following types are added to this list:
+
+- REQUEST/RESPONSE
+- ERROR
+
+For the `REQUEST` frame, the following `type`s are defined:
+
+- SWAP
```
Comment (LLFourn) [inline RFC-001-BAM.md]: :x: this should be in RFC002

File: RFC-002-SWAP.md [hunk 35]
```diff
@@ -10,267 +10,255 @@
...
+Subsequent RFCs can refer to this type if they want to define a particular asset.
+
+### The type `SwapProtocol`
+
+A section "Protocols" is added to the registry which tracks all currently defined protocols.
+Subsequent RFCs can refer to this type if they want to define a particular swap protocol.
+
+### The type `DeclineBody` and the initial set of possible values
+
+A section "DeclineBodys" is added to the registry which tracks all currently defined reasons.
```
Comment (LLFourn) [inline RFC-002-SWAP.md]:
```suggestion
A section "DeclineBody" is added to the registry which tracks all currently defined reasons.
```

File: assets/RFC002-parse-response-body.puml (hunk 32 above)
[the remark this answers is not in the document]
Comment (D4nte) [reply assets/RFC002-parse-response-body.puml]: > The rendered version is embedded in the RFC once it is merged and available on master.

Shame we cannot review the rendered version

File: RFC-002-SWAP.md [hunk 36]
```diff
@@ -10,267 +10,255 @@
...
+
+#### `protocol-unsupported`
+
+The protocol specified in the `protocol` header is not known to the receiving party.
+
+##### `details`
+
+TBD <!-- List known protocols in details -->
+
+#### `unknown-ledger`
```
[the remark this answers is not in the document]
Comment (thomaseizinger) [reply RFC-002-SWAP.md]: Interesting idea.

`unknown-ledger` could be something that is generated by a deserializer (hence automatic) whereas `unsupported` would mean a configuration somewhere.

I am leaning towards having that separated because (1) means there is no point in retrying and (2) might be successful later.

File: RFC-001-BAM.md [hunk 37]
```diff
@@ -166,127 +173,112 @@ Ids MAY be skipped but MUST be ascending.
 
 > `NOTIFICATION` is a reserved type and may be specified at a later point.
 
-#### Request / Response
+#### Error
 
-A request is a type of message that implies an answer.
-Nodes MUST be prepared to receive a frame of type `RESPONSE` with the id used in the `REQUEST` frame.
+The `ERROR` frame is used to communicate failures between nodes at the communication level.
+They MUST NOT be used to communicate errors on the application level.
```
[the remark this answers is not in the document]
Comment (thomaseizinger) [reply RFC-001-BAM.md]: I would say those need to go into a `SWAP RESPONSE` with the `decision` header set to `declined` and an appropriate body.

File: RFC-001-BAM.md (hunk 34 above)
Comment (thomaseizinger) [reply RFC-001-BAM.md]: Oh stupid me, of course :D

File: .vscode/settings.json [hunk 38]
```diff
@@ -4,7 +4,7 @@
     "editor.tabSize": 4,
     "markdown.extension.list.indentationSize": "inherit",
     "spellright.language": [
-        "en_US"
+        "en_AU"
```
[the remark this answers is not in the document]
Comment (thomaseizinger) [reply .vscode/settings.json]: It is actually not working so I will change it again :D

File: RFC-001-BAM.md [hunk 39]
```diff
@@ -166,127 +173,112 @@ Ids MAY be skipped but MUST be ascending.
 
 > `NOTIFICATION` is a reserved type and may be specified at a later point.
 
-#### Request / Response
+#### Error
```
[the remark this answers is not in the document]
Comment (thomaseizinger) [reply RFC-001-BAM.md]: Makes sense, will move it up!

File: RFC-003-SWAP-Basic.md [hunk 40]
```diff
@@ -103,39 +105,24 @@ How to construct HTLCs based on SHA-256 and other hash functions for particular
 ### SWAP Request Body
```
[the remark this answers is not in the document]
Comment (thomaseizinger) [reply RFC-003-SWAP-Basic.md]: Please open a separate issue for this :)

File: assets/RFC002-parse-response-body.puml (hunk 32 above)
[the remark this answers is not in the document]
Comment (thomaseizinger) [reply assets/RFC002-parse-response-body.puml]: You are free to just copy/paste it into plantuml online: http://plantuml.com/

File: RFC-002-SWAP.md (hunk 36 above)
[the remark this answers is not in the document]
Comment (bonomat) [reply RFC-002-SWAP.md]: For me this was more of a semantical issue, i.e. `unsupported-ledger` implies `unknown-ledger` while the other way around does not hold, e.g. we could have a generic ledger implementation which can handle any ledger (🕺)

[the remark this answers is not in the document]
Comment (D4nte) [reply RFC-002-SWAP.md]: The separation sounds good but I do not see any `unsupported-*` in the file. Please add it.

File: RFC-001-BAM.md (hunk 37 above)
[the remark this answers is not in the document]
Comment (LLFourn) [reply RFC-001-BAM.md]: Note: Discussed with @thomaseizinger. We'll roll with this. The drawback of this approach is that when implementing a `RESPONSE` type message you have to roll your own error for problems with headers. Maybe reconsider this in next revision.
